lib/restbase.js

```diff
@@ -173,79 +173,135 @@ RESTBase.prototype.defaultWebRequestHandler = function(req) {
     if (this._recursionDepth === 0 &&
-            ((req.uri.params && req.uri.params.api === 'sys')
-             // TODO: Remove once params.api is reliable
-             || (req.uri.path.length > 1 && req.uri.path[1] === 'sys'))) {
-        return P.reject(new HTTPError({
+            (match && match.params && match.params.api === 'sys'
+            // Fallback to protect against accidentally exposing system routes in RESTBase
+            || req.uri.path.length > 1 && req.uri.path[1] === 'sys')) {
+        throw new HTTPError({
             status: 403,
             body: {
                 type: 'access_denied#sys',
-                title: 'Access to the /{domain}/sys/ hierarchy is restricted to system users.'
+                title: 'Access to the /sys hierarchy is restricted to system users.'
             }
-        }));
+        });
     }
+};
 
+RESTBase.prototype.request = function(req) {
     if (req.method) {
         req.method = req.method.toLowerCase();
     }
+    return P.try(this._request.bind(this, req));
+};
 
-    return this._request(req);
+RESTBase.prototype._wrapInMetrics = function(handlerPromise, match, req) {
+    var self = this;
+    // Remove the /{domain}/ prefix, as it's not very useful in stats
+    var statName = match.value.path.replace(/\/[^\/]+\//, '')
+            + '.' + req.method.toUpperCase() + '.';
+    // Normalize invalid chars
+    statName = self.metrics.normalizeName(statName);
+    // Start timer
+    var startTime = Date.now();
+
+    return handlerPromise.then(function(res) {
+        // Record request metrics & log
+        var statusClass = Math.floor(res.status / 100) + 'xx';
+        self.metrics.endTiming([statName + statusClass, statName + 'ALL'], startTime);
+        return res;
+    },
+    function(err) {
+        var statusClass = '5xx';
+        if (err && err.status) {
+            statusClass = Math.floor(err.status / 100) + 'xx';
+        }
+        self.metrics.endTiming([statName + statusClass, statName + 'ALL'], startTime);
+        throw err;
+    });
 };
 
+RESTBase.prototype._wrapInAccessCheck = function(handlerPromise, match, childReq) {
+    var self = this;
+    if (match.permissions && Array.isArray(match.permissions) && match.permissions.length) {
+        self._authService = self._authService || new AuthService(match.value.specRoot);
+        self._authService.addRequirements(match.permissions);
+        if (childReq.method === 'get' || childReq.method === 'head') {
+            return P.all([
+                handlerPromise,
+                self._authService.checkPermissions(self, childReq)
+            ])
+            .then(function(res) { return res[0]; });
+        } else {
+            return self._authService.checkPermissions(self, childReq)
+            .then(function() { return handlerPromise; });
+        }
+    } else {
+        return handlerPromise;
+    }
+};
 
-// Internal request handler
+// Process one request
```

Comment:
Since this code path it's very hot it would be good to double-check that there is no significant performance penalty from this refactor. It seems unlikely that it would have a big effect, but good to check.